// @admin-bro/design-system/build/molecules/modal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Modal = void 0;
const react_1 = __importDefault(require("react"));
const styled_components_1 = __importStar(require("styled-components"));
const css_class_1 = require("../utils/css-class");
const __1 = require("..");
const ModalBox = styled_components_1.default.div `
  margin: 30px auto;
  max-width: 540px;
  max-height: 90vh;
  padding: 16px;
  padding-left: 32px;
  background-color: ${({ theme }) => theme.colors.white};

  ${({ warning }) => warning && styled_components_1.css `
    border-left: 8px solid ${({ theme }) => theme.colors.error};
  `}
`;
const CloseSection = styled_components_1.default.div `
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: center;
`;
const CloseButton = styled_components_1.default(__1.Button) `
  border: none;
`;
const ModalSection = styled_components_1.default.div `
  padding: 15px 15px 15px 0;
`;
/**
 * Modal box, shows informations, has warning label
 *
 * Usage:
 * ```javascript
 * import { Section } from '@admin-bro/design-system'
 * ```
 * @component Modal
 * @subcategory Atoms
 * @example
 * return (
 *   <Modal open={open} warning={warning}>
 *    <Text fontSize={16} fontWeight="600">Are you sure you want to delete this car?</Text>
 *    <Text fontSize={14}>This item will be deleted immediately. You can’t undo this action.</Text>
 *    <Buttons>
 *      <Button size="sm" variant="secondary" marginRight={15}>Cancel</Button>
 *      <Button size="sm" variant="primary">Delete</Button>
 *    </Buttons>
 *  </Modal>
 * )
 * @memberof module:@admin-bro/design-system
 */
const Modal = (_a) => {
    var { open, onCloseClick, warning, className, children } = _a, other = __rest(_a, ["open", "onCloseClick", "warning", "className", "children"]);
    if (!open) {
        return null;
    }
    return (react_1.default.createElement(__1.Box, { variant: "grey", height: "100vh" },
        react_1.default.createElement(ModalBox, Object.assign({ warning: warning, className: className }, other),
            react_1.default.createElement(CloseSection, null,
                react_1.default.createElement("span", null, warning && (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(__1.Icon, { icon: "Warning", color: "error" }),
                    react_1.default.createElement(__1.Label, { inline: true, pl: "sm", color: "error" }, "Warning")))),
                react_1.default.createElement(CloseButton, { rounded: true, size: "icon", onClick: onCloseClick },
                    react_1.default.createElement(__1.Icon, { icon: "CloseOutline", color: "primary" }))),
            react_1.default.createElement(ModalSection, null, children))));
};
exports.Modal = Modal;
Modal.defaultProps = {
    className: css_class_1.cssClass('Section'),
};
exports.default = Modal;
