// @admin-bro/design-system/build/atoms/badge.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Default = void 0;
const addon_knobs_1 = require("@storybook/addon-knobs");
const react_1 = __importDefault(require("react"));
const badge_1 = __importDefault(require("./badge"));
exports.default = { title: 'DesignSystem/Atoms/Badge', decorators: [addon_knobs_1.withKnobs] };
var BadgeVariant;
(function (BadgeVariant) {
    BadgeVariant["Primary"] = "primary";
    BadgeVariant["Danger"] = "danger";
    BadgeVariant["Text"] = "text";
    BadgeVariant["Success"] = "success";
    BadgeVariant["Info"] = "info";
    BadgeVariant["Secondary"] = "secondary";
})(BadgeVariant || (BadgeVariant = {}));
var BadgeSize;
(function (BadgeSize) {
    BadgeSize["Sm"] = "sm";
    BadgeSize["Lg"] = "lg";
})(BadgeSize || (BadgeSize = {}));
exports.Default = () => {
    const variant = addon_knobs_1.select('Variant', Object.values(BadgeVariant), BadgeVariant.Primary);
    const size = addon_knobs_1.select('Size', Object.values(BadgeSize), BadgeSize.Sm);
    const outline = addon_knobs_1.boolean('Outline', false);
    return (react_1.default.createElement(badge_1.default, { variant: variant, size: size, outline: outline }, `${variant} ${size} ${outline ? 'outline' : ''}`));
};
