// @admin-bro/design-system/build/atoms/badge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Badge = void 0;
const styled_components_1 = __importDefault(require("styled-components"));
const styled_system_1 = require("styled-system");
const css_class_1 = require("../utils/css-class");
const variantStyle = (variantColor, props) => ({
    bg: variantColor,
    borderColor: variantColor,
    color: props.outline ? variantColor : 'white',
});
const colorVariant = (props) => styled_system_1.variant({
    variants: {
        primary: variantStyle('primary100', props),
        danger: variantStyle('error', props),
        success: variantStyle('success', props),
        info: variantStyle('info', props),
        secondary: variantStyle('accent', props),
    },
});
const sizeVariants = styled_system_1.variant({
    prop: 'size',
    variants: {
        sm: {
            py: 'xs',
        },
        lg: {
            py: 'default',
            px: '10px',
        },
    },
});
/**
 * Component representing a badge.
 *
 * Usage
 * ```javascript
 * import { Badge, BadgeProps } from '@admin-bro/design-system'
 * ```
 *
 * @component
 * @subcategory Atoms
 * @example <caption>Color variants</caption>
 * const variants = ['primary', 'danger', 'success', 'info', 'secondary']
 * return (
 * <Box py="lg">
 *   <Badge mb="default" mr="default">default</Badge>
 *   {variants.map(variant => (
 *     <Badge mb="default" variant={variant} mr="default">{variant}</Badge>
 *   ))}
 * </Box>
 * )
 * @example <caption>Outline badges</caption>
 * const variants = ['primary', 'danger', 'success', 'info', 'secondary']
 * return (
 * <Box py="lg">
 *   <Badge mb="default" mr="default" outline>default</Badge>
 *   {variants.map(variant => (
 *     <Badge mb="default" variant={variant} mr="default" outline>{variant}</Badge>
 *   ))}
 * </Box>
 * )
 * @example <caption>Different sizes</caption>
 * return (
 * <Box py="lg">
 *   <Badge ml="default" variant="primary" size="sm">small</Badge>
 *   <Badge ml="default" variant="primary">regular</Badge>
 *   <Badge ml="default" variant="primary" size="lg">large</Badge>
 * </Box>
 * )
 * @memberof module:@admin-bro/design-system
 */
const Badge = styled_components_1.default.span `
  border-radius: 12px;
  border: 1px solid ${({ theme }) => theme.colors.grey40};
  color: ${({ outline, theme }) => (outline ? theme.colors.grey60 : theme.colors.white)};
  vertical-align: middle;
  font-family: ${({ theme }) => theme.font};

  ${styled_system_1.space};
  ${styled_system_1.color};
  ${styled_system_1.typography};
  ${(props) => colorVariant(props)};
  ${sizeVariants};
  ${({ outline }) => (outline ? 'background: transparent;' : '')}
`;
exports.Badge = Badge;
Badge.defaultProps = {
    px: 'default',
    py: '6px',
    fontSize: 'xs',
    bg: 'grey40',
    className: css_class_1.cssClass('Badge'),
};
exports.default = Badge;
