// @admin-bro/design-system/build/atoms/button.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Default = void 0;
const addon_knobs_1 = require("@storybook/addon-knobs");
const react_1 = __importDefault(require("react"));
const button_1 = __importDefault(require("./button"));
exports.default = { title: 'DesignSystem/Atoms/Button', decorators: [addon_knobs_1.withKnobs] };
var ButtonVariant;
(function (ButtonVariant) {
    ButtonVariant["Primary"] = "primary";
    ButtonVariant["Danger"] = "danger";
    ButtonVariant["Text"] = "text";
    ButtonVariant["Success"] = "success";
    ButtonVariant["Info"] = "info";
    ButtonVariant["Secondary"] = "secondary";
})(ButtonVariant || (ButtonVariant = {}));
var ButtonSize;
(function (ButtonSize) {
    ButtonSize["Small"] = "sm";
    ButtonSize["Large"] = "lg";
    ButtonSize["Icon"] = "icon";
})(ButtonSize || (ButtonSize = {}));
exports.Default = () => {
    const variant = addon_knobs_1.select('Variant', Object.values(ButtonVariant), ButtonVariant.Primary);
    const size = addon_knobs_1.select('Size', Object.values(ButtonSize), ButtonSize.Small);
    const rounded = addon_knobs_1.boolean('Rounded', false);
    const disabled = addon_knobs_1.boolean('Disabled', false);
    return (react_1.default.createElement(button_1.default, { variant: variant, size: size, rounded: rounded, disabled: disabled }, `${variant} ${size} ${rounded ? 'rounded' : ''} ${disabled ? 'disabled' : ''}`));
};
