// @admin-bro/design-system/build/atoms/header.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.H6 = exports.H5 = exports.H4 = exports.H3 = exports.H2 = exports.H1 = exports.Header = exports.default = void 0;
const react_1 = __importDefault(require("react"));
const styled_components_1 = __importDefault(require("styled-components"));
const styled_system_1 = require("styled-system");
const button_1 = require("./button");
const css_class_1 = require("../utils/css-class");
const Base = styled_components_1.default.h3 `
  font-family: ${({ theme }) => theme.font};
  vertical-align: middle;
  margin: ${({ theme }) => theme.space.default} 0;
  padding: 0;
  * {
    vertical-align: middle;
  }
  & ${button_1.Button}, a {
    vertical-align: bottom;
  }
  ${styled_system_1.typography};
  ${styled_system_1.space};
`;
Base.defaultProps = {
    fontWeight: 'normal',
    fontSize: 'h3',
    lineHeight: 'xl',
    className: css_class_1.cssClass(['Header', 'H3']),
};
const H1 = styled_components_1.default((props) => react_1.default.createElement(Base, Object.assign({ as: "h1" }, props))) ``;
exports.H1 = H1;
H1.defaultProps = {
    fontSize: 'h1',
    lineHeight: 'xxl',
    className: css_class_1.cssClass(['Header', 'H1']),
};
const H2 = styled_components_1.default((props) => react_1.default.createElement(Base, Object.assign({ as: "h2" }, props))) `
  & ${button_1.Button}, a {
    margin-bottom: 4px;
  }
`;
exports.H2 = H2;
H2.defaultProps = {
    fontSize: 'h2',
    lineHeight: 'xxl',
    className: css_class_1.cssClass(['Header', 'H2']),
};
const H3 = Base;
exports.H3 = H3;
const H4 = styled_components_1.default((props) => react_1.default.createElement(Base, Object.assign({ as: "h4" }, props))) ``;
exports.H4 = H4;
H4.defaultProps = {
    fontSize: 'h4',
    lineHeight: 'xl',
    className: css_class_1.cssClass(['Header', 'H4']),
};
const H5 = styled_components_1.default((props) => react_1.default.createElement(Base, Object.assign({ as: "h5" }, props))) ``;
exports.H5 = H5;
H5.defaultProps = {
    fontSize: 'xl',
    lineHeight: 'lg',
    className: css_class_1.cssClass(['Header', 'H5']),
};
const H6 = styled_components_1.default((props) => react_1.default.createElement(Base, Object.assign({ as: "h6" }, props))) ``;
exports.H6 = H6;
H6.defaultProps = {
    fontSize: 'lg',
    lineHeight: 'lg',
    className: css_class_1.cssClass(['Header', 'H6']),
};
/**
 *
 * The Header component is a base for all text components intended as headings.
 *
 * Usage
 * ```javascript
 * import { H1, H2, H3, H4, H5, H6, HeaderProps } from '@admin-bro/design-system'
 *
 * // or
 *
 * import { Header } from '@admin-bro/design-system'
 * // Header.H1
 * ```
 * By default, the Heading component renders an __h3__ element.
 *
 * @component
 * @subcategory Atoms
 * @example
 * return (
 * <Box py="lg">
 *   <Header.H1>H1 Header - 40</Header.H1>
 *   <Text variant="sm" mb={5}>Roboto 40 - line height - 40</Text>
 *   <Header.H2>H2 Header - 32</Header.H2>
 *   <Text variant="sm" mb={5}>Roboto 32 - line height - 40</Text>
 *   <Header.H3>H3 Header - 28</Header.H3>
 *   <Text variant="sm" mb={5}>Roboto 28 - line height - 32</Text>
 *   <Header.H4>H4 Header - 24</Header.H4>
 *   <Text variant="sm" mb={5}>Roboto 24 - line height - 32</Text>
 *   <Header.H5>H5 Header - 18</Header.H5>
 *   <Text variant="sm" mb={5}>Roboto 18 - line height - 24</Text>
 *   <Header.H6>H6 Header - 16</Header.H6>
 *   <Text variant="sm" mb={5}>Roboto 16 - line height - 24</Text>
 * </Box>
 * )
 *
 * @memberof module:@admin-bro/design-system
 */
const Header = H3;
exports.default = Header;
exports.Header = Header;
Header.H1 = H1;
Header.H2 = H2;
Header.H3 = H3;
Header.H4 = H4;
Header.H5 = H5;
Header.H6 = H6;
