// @admin-bro/design-system/build/atoms/link.stories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Default = void 0;
const addon_knobs_1 = require("@storybook/addon-knobs");
const react_1 = __importDefault(require("react"));
const link_1 = __importDefault(require("./link"));
exports.default = { title: 'DesignSystem/Atoms/Link', decorators: [addon_knobs_1.withKnobs] };
var LinkVariant;
(function (LinkVariant) {
    LinkVariant["Primary"] = "primary";
    LinkVariant["Danger"] = "danger";
    LinkVariant["Success"] = "success";
    LinkVariant["Info"] = "info";
    LinkVariant["Secondary"] = "secondary";
})(LinkVariant || (LinkVariant = {}));
var LinkSize;
(function (LinkSize) {
    LinkSize["Sm"] = "sm";
    LinkSize["Lg"] = "lg";
})(LinkSize || (LinkSize = {}));
exports.Default = () => {
    const uppercase = addon_knobs_1.boolean('Defines if link should be uppercase', false);
    const variant = addon_knobs_1.select('Color variant', Object.values(LinkVariant), LinkVariant.Primary);
    const size = addon_knobs_1.select('Size variant', Object.values(LinkSize), LinkSize.Sm);
    return (react_1.default.createElement(link_1.default, { uppercase: uppercase, variant: variant, size: size, href: "/" }, `just ${variant} ${size} ${uppercase ? 'uppercased' : ''} link`));
};
