// @admin-bro/design-system/build/atoms/table/table-caption.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const styled_components_1 = __importDefault(require("styled-components"));
const button_1 = require("../button");
const text_1 = require("../text");
const css_class_1 = require("../../utils/css-class");
const CAPTION_HEIGHT = '42px';
/**
 * @component
 * @private
 */
const TableCaption = styled_components_1.default.caption `
  font-family: ${({ theme }) => theme.font};
  padding: ${({ theme }) => theme.space.sm} ${({ theme }) => theme.space.lg};
  text-align: left;
  color: ${({ theme }) => theme.colors.white};
  font-size: ${({ theme }) => theme.fontSizes.default};
  line-height: ${({ theme }) => theme.lineHeights.default};
  position: absolute;
  height: ${CAPTION_HEIGHT};
  left: 0;
  right: 0;
  top: -${CAPTION_HEIGHT};
  background: ${({ theme }) => theme.colors.primary100};
  box-sizing: border-box;

  & ${button_1.Button} {
    color: ${({ theme }) => theme.colors.white};
    & > span svg {
      fill: ${({ theme }) => theme.colors.white};
    }
    &:hover {
      color: ${({ theme }) => theme.colors.white};
    }
  }

  & ${text_1.Text} {
    color: ${({ theme }) => theme.colors.white};
  }
`;
TableCaption.defaultProps = {
    className: css_class_1.cssClass('TableCaption'),
};
exports.default = TableCaption;
