// @admin-bro/design-system/build/molecules/drop-down/drop-down-item.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropDownItem = void 0;
const styled_components_1 = __importDefault(require("styled-components"));
const styled_system_1 = require("styled-system");
const link_1 = __importDefault(require("../../atoms/link"));
const box_1 = __importDefault(require("../../atoms/box"));
/**
 * @component
 * @private
 */
exports.DropDownItem = styled_components_1.default(box_1.default) `
  position: relative;
  z-index: 10000;
  border: none;
  color: ${({ theme }) => theme.colors.grey80};
  display: block;
  font-family: ${({ theme }) => theme.font};
  border: solid transparent;
  border-width: 0 ${({ theme }) => theme.space.sm};
  &:hover {
    border-color: ${({ theme }) => theme.colors.primary100};
    background: ${({ theme }) => theme.colors.grey20};
  }
  & svg {
    vertical-align: middle;
    padding-bottom: 2px;
    padding-right: ${({ theme }) => theme.space.default};
    fill: ${({ theme }) => theme.colors.grey40};
  }
  & a {
    color: ${({ theme }) => theme.colors.grey80};
  }

  ${styled_system_1.space};

  & > ${link_1.default}, & > a {
    padding: ${({ theme }) => theme.space.lg};
    display: block;
    &:hover{
      text-decoration: none;
    }
  }
`;
exports.default = exports.DropDownItem;
