// @admin-bro/design-system/build/templates/navigation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Navigation = void 0;
/* eslint-disable import/prefer-default-export */
const styled_components_1 = __importDefault(require("styled-components"));
const box_1 = require("../atoms/box");
const css_class_1 = require("../utils/css-class");
exports.Navigation = styled_components_1.default(box_1.Box) `
  height: 100%;
  width: ${({ theme }) => theme.sizes.sidebarWidth};
  padding: ${({ theme }) => theme.space.lg};
  display: flex;
  flex-direction: column;
  overflow: auto;
  border-right: 1px solid ${({ theme }) => theme.colors.grey20};
  flex-shrink: 0;
  background: ${({ theme }) => theme.colors.white};
  z-index: 50;
  transition: all 500ms;
  left: 0;

  &.hidden {
    left: -${({ theme }) => theme.sizes.sidebarWidth};
    transition: all 500ms;
  }
`;
exports.Navigation.defaultProps = {
    className: css_class_1.cssClass('Navigation'),
};
